src/lib/Optimize.hs

```diff
@@ -211,6 +211,92 @@ instance GenericTraverser ULS where
 emitSubstBlock :: Emits o => Block i -> ULM i o (Atom o)
 emitSubstBlock (Block _ decls ans) = traverseDeclNest decls $ traverseAtom ans
 
+-- === Loop invariant code motion ===
+
+-- TODO: Resolve import cycle with Lower
+type IxDestBlock = Abs (Nest Decl) (Abs Binder Block)
+
+hoistLoopInvariant :: EnvReader m => IxDestBlock n -> m n (IxDestBlock n)
+hoistLoopInvariant (Abs ixs (Abs (db:>dTy) body)) =
+  liftM fst $ liftGenericTraverserM LICMS $
+    buildScoped $ traverseDeclNest ixs do
+      dTy' <- traverseGenericE dTy
+      buildAbs (getNameHint db) dTy' \v -> extendRenamer (db@>v) $ traverseGenericE body
+
+data LICMS (n::S) = LICMS
+instance SinkableE LICMS where
+  sinkingProofE _ LICMS = LICMS
+instance HoistableState LICMS where
+  hoistState _ _ LICMS = LICMS
+
+instance GenericTraverser LICMS where
+  traverseExpr = \case
+    Hof (Seq dir ix (ProdVal dests) (Lam (LamExpr b body@(Block ann _ _)))) -> do
+      Distinct <- getDistinct
+      ix' <- traverseAtom ix
+      dests' <- traverse traverseAtom dests
+      let numCarries = length dests
+      Abs hdecls destsAndBody <- traverseLamBinder b \b' -> do
+        -- First, traverse the block, to allow any Hofs inside it to hoist their own decls.
+        Block _ decls ans <- traverseGenericE body
```

Comment:
Is this going to be quadratic in the loop nesting depth? Not a blocking problem if so. Just curious.

Reply:
I think it's still linear. This `traverseGenericE` is the first traversal of the body, while the cost of the `seqLICM` call below is equal to the number of decls _in the returned block_. Any decls nested in blocks inside are not re-traversed, since we never open up the expressions.

Well, I guess for as long as we don't hoist the allocs and the substitution is the identity substitution... Otherwise I think we do need to re-traverse to substitute and that is potentially exponential in the nesting depth...